WordPress/gutenberg
# Initial POC for Navigation Overlay as template part

## What?

Updates the navigation block to use a template part for the overlay.

## Why?

Background: https://github.com/WordPress/gutenberg/issues/43852

## How?
**WIP: This is still very buggy.**
In the first attempt, the overlay template part id was stored as a block attribute.

Now the block tries to find a template part that matches the current menu, and it is not working yet:
- When you change the menu in the menu selection list, the new menu id is not passed to the overlay panel until the page is refreshed.
- The front view is **not** implemented.

<details> <summary>Previous description </summary>
The PR adds a new Navigation Overlay selection option to the block settings sidebar. This option is similar to the menu selection.
It is a select list that shows the available navigation overlay template parts and lets you select, create a new overlay, and edit the current overlay.
When the option to Edit Overlay is selected, the template part is opened in isolation, and any block can be added.

On the front the template part is shown when the responsive menu button is activated.

## Why?

Background: https://github.com/WordPress/gutenberg/issues/43852

## How?

In short:
Adds a new block attribute that holds the postId of the assigned template part.
Selects all `wp_template_part` post types from the database, and filters the result to only include the navigation overlays.
When an overlay is selected from the setting, the block attribute is updated.
The attribute is used to direct the user to the correct template part for editing,
and also for rendering the part on the front.

## Testing Instructions

1. Activate a block theme. This will not work on classic themes without template part editing.
4. In either editor, select or add a navigation block.
5. Find the Overlay panel in the block settings sidebar. Previously the overlay settings were in a panel called Display. This panel has been replaced by the separate Overlay and Submenu panels.
6. Set the Overlay to Mobile.
7. Next please select the "Navigation Overlay" button.
8. A tools modal should open where you can select the option "Create new Overlay".
9. **When selecting this option you are probably expecting to be redirected to where you can edit the overlay. Leave a comment if you know how to do this in a non hacky way 😃** 
10. Confirm that after selecting the "Create new Overlay" option, you have unsaved changes for the navigation block. Save.
11. Below the select option there should now be a new link with the link text Edit Overlay.
12. Activate this link. You should now be in the template editor with the Navigation Overlay opened in edit mode. Make some changes to the default blocks or add some new blocks and style choices. Save
13. **Now you may want to return to the template or post you were just editing... let me know if you know how to do this.**
14. Go back to the template where you added the navigation block. Set the Overlay to Always.
15. The responsive menu button should now be visible.
In the issue description, clicking this button should take you to the template editor.
**Bug:** This is currently not working _because it tries to open the site editor inside the current editor iframe._  
I also found it more difficult to select the block in the canvas when this button is set to open the template editor.

_Phew._ 
1. Now please view the navigation block on the front.
4. Activate the responsive menu button.
5. You should see the blocks you added to the template part below the close button.
6. **Bug:** You will also find that it doesn't look like you expected, as the template part is rendered after the CSS has already been added.
**Bug:** If you tested the mobile menu overlay, you will also see that the template part is always displayed, even on full browser width.
</detail>
## Screenshots or screencast 

https://user-images.githubusercontent.com/7422055/212088670-2ad9f0b8-f287-40b2-95d0-07c6e518f2cd.mov
